casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala

```diff
@@ -0,0 +1,74 @@
+package coop.rchain.casper
+
+import cats.Monad
+import cats.effect.{Concurrent, Sync}
+import cats.implicits._
+import coop.rchain.blockstorage.{BlockDagRepresentation, BlockDagStorage, BlockStore}
+import coop.rchain.casper.util.ProtoUtil
+import coop.rchain.shared.{Cell, Log}
+import coop.rchain.catscontrib.ListContrib
+
+object LastFinalizedBlockCalculator {
+  // TODO: Extract hardcoded fault tolerance threshold
+  private val faultToleranceThreshold = 0f
+
+  def run[F[_]: Monad: Log: Sync: Concurrent: BlockStore: BlockDagStorage: SafetyOracle](
+      dag: BlockDagRepresentation[F],
+      lastFinalizedBlockHash: BlockHash
+  )(implicit state: Cell[F, CasperState]): F[BlockHash] =
+    for {
+      maybeChildrenHashes <- dag.children(lastFinalizedBlockHash)
+      childrenHashes      = maybeChildrenHashes.getOrElse(Set.empty[BlockHash]).toList
+      maybeFinalizedChild <- ListContrib.findM(
+                              childrenHashes,
+                              (blockHash: BlockHash) =>
+                                isGreaterThanFaultToleranceThreshold(dag, blockHash)
+                            )
+      newFinalizedBlock <- maybeFinalizedChild match {
+                            case Some(finalizedChild) =>
+                              removeDeploysInFinalizedBlock[F](finalizedChild) *> run[F](
+                                dag,
+                                finalizedChild
+                              )
+                            case None => lastFinalizedBlockHash.pure[F]
+                          }
+    } yield newFinalizedBlock
+
+  private def removeDeploysInFinalizedBlock[F[_]: Monad: Log: Sync: Concurrent: BlockStore: BlockDagStorage: SafetyOracle](
+      finalizedChildHash: BlockHash
+  )(implicit state: Cell[F, CasperState]): F[Unit] =
+    for {
+      block              <- ProtoUtil.unsafeGetBlock[F](finalizedChildHash)
+      deploys            = block.body.get.deploys.map(_.deploy.get).toList
+      stateBefore        <- state.read
+      initialHistorySize = stateBefore.deployHistory.size
+      _ <- state.modify { s =>
+            s.copy(deployHistory = s.deployHistory -- deploys)
+          }
+      stateAfter     <- state.read
+      deploysRemoved = initialHistorySize - stateAfter.deployHistory.size
+      _ <- Log[F].info(
+            s"Removed $deploysRemoved deploys from deploy history as we finalized block ${PrettyPrinter
+              .buildString(finalizedChildHash)}."
+          )
+    } yield ()
+
+  /*
+   * On the first pass, block B is finalized if B's main parent block is finalized
```

Comment:
This comment looks outdated. The method it applies to does not look like it's checking whether the main parent block is finalized

Reply:
LastFinalizedBlockCalculator.run is only recursively called if there was a block that was finalized.

Reply:
You mean if the "main parent block" was finalized? Or any?

Reply:
Main parent block